WFLY-13147 Avoid processing deployments multiple times in Weld subsys…

manovotn opened the pull request:
…tem.

JIRA - https://issues.redhat.com/browse/WFLY-13147

_This basically a copy of what the issue suggsested as code changes_. All in all those changes make sense in avoiding loading dep modules that we know were/will be processed as separate units. Simple tests pass for me (as in tests in this repo executed with `-DallTests`).

However, I don't have a big enough sample to actually measure the impact on re-deploy times on a kind of application that it was reported against - @bstansberry does WFLY have some perf tests for deployment where this could be measured?
Or really, any more "obscure" deployment tests that we could/should run against.

CCing @mkouba just in case he has some thoughts about these changes.

manovotn committed:
WFLY-13147 Avoid processing deployments multiple times in Weld subsystem.

bstansberry commented:
@manovotn WildFly as a project doesn't have separate boot perf tests. Our testsuite plus the EE TCKs is basically what we have. @msve do you guys have anything like that?

@ropalka I know last you you did some work on the number of services EAP deployments were creating. If you have some sort of deployment speed test left over from that that might be useful, please let @manovotn  know.

bstansberry commented on weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/ExternalBeanArchiveProcessor.java:
```diff
@@ -162,6 +162,14 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
+                    continue;
+                }
+                if (deploymentUnits.stream().anyMatch(d -> (((ModuleDependencySpec) dep).getName()).endsWith(d.getName()))) {
+                    // d.getName() returns deployment.123abc.ear, whereas dep.getName() returns 123abc.ear
```
This comment is confusing because for using the example values the code would amount to "123abc.ear".endsWith("deployment.123abc.ear") which would never be true.

manovotn replied:
Oh yea, it's the other way around of course, my bad. Either way, I wasn't planning to keep it here, it was just a temporary explanation of why the matching works based on my debugging...

ropalka commented:
Hi @bstansberry @manovotn - I didn't measure the boot time when I was dealing with number of services. What I have been doing was I was just counting number of created services during the deployment phase.

ucswntr commented:
Hi @bstansberry @manovotn , are there any news regarding the patch? Our colleagues are already  looking forward to reducing the startup times during their development/testing cycles.

KR Christoph

manovotn commented:
I know about this issue but I keep being pulled to other issues and PRs.

If @bstansberry doesn't have anything against it, we can let it in like this. The changes don't break existing tests and we know that they lower deployment times for @ucswntr scenario. 

I've fixed the misleading comment and rebased the PR against master.

bstansberry commented on weld/subsystem/src/main/java/org/jboss/as/weld/deployment/processors/ExternalBeanArchiveProcessor.java:
```diff
@@ -162,6 +162,14 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
+                if (!(dep instanceof ModuleDependencySpec)) {
+                    continue;
+                }
+                if (deploymentUnits.stream().anyMatch(d -> (((ModuleDependencySpec) dep).getName()).endsWith(d.getName()))) {
```
My only "maybe" suggestion is earlier in this method when looping through the DeploymentUnits, built up a Set<String> containing "deployment." + d.getName().  Then here just do a contains check against that set.

bstansberry commented:
@manovotn I'm fine with adding this. I made a small comment above but if you don't like that, please just @ ping me and I'll approve this and it can go in.

manovotn commented:
@bstansberry updated, see if you like it now :-)
